src/main/java/software/amazon/awssdk/s3accessgrants/cache/S3AccessGrantsCachedBucketRegionResolver.java

```diff
@@ -0,0 +1,170 @@
+package software.amazon.awssdk.s3accessgrants.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import software.amazon.awssdk.core.exception.SdkServiceException;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.HeadBucketRequest;
+import software.amazon.awssdk.services.s3.model.HeadBucketResponse;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+import software.amazon.awssdk.utils.Logger;
+
+import java.time.Duration;
+
+import static software.amazon.awssdk.s3accessgrants.cache.S3AccessGrantsConstants.BUCKET_REGION_CACHE_SIZE;
+import static software.amazon.awssdk.s3accessgrants.cache.S3AccessGrantsConstants.BUCKET_REGION_EXPIRE_CACHE_AFTER_WRITE_SECONDS;
+import static software.amazon.awssdk.s3accessgrants.cache.S3AccessGrantsConstants.MAX_BUCKET_REGION_EXPIRE_CACHE_AFTER_WRITE_SECONDS;
+import static software.amazon.awssdk.s3accessgrants.cache.S3AccessGrantsConstants.MAX_BUCKET_REGION_CACHE_SIZE;
+
+public class S3AccessGrantsCachedBucketRegionResolver implements S3AccessGrantsBucketRegionResolver{
+
+
+    private int maxCacheSize;
+
+    private int expireCacheAfterWriteSeconds;
+
+    private static Cache<String, Region> cache;
+
+    public int getMaxCacheSize() {
+        return maxCacheSize;
+    }
+
+
+    public int getExpireCacheAfterWriteSeconds() {
+        return expireCacheAfterWriteSeconds;
+    }
+
+    private S3Client s3Client;
+
+    private static final Logger logger = Logger.loggerFor(S3AccessGrantsCachedBucketRegionResolver.class);
+
+    public int maxCacheSize() {
+        return maxCacheSize;
+    }
+
+    public int expireCacheAfterWriteSeconds() {
+        return expireCacheAfterWriteSeconds;
+    }
+
+    protected CacheStats getCacheStats() { return cache.stats(); }
+
+    S3AccessGrantsCachedBucketRegionResolver() {
+        this.maxCacheSize = BUCKET_REGION_CACHE_SIZE;
+        this.expireCacheAfterWriteSeconds = BUCKET_REGION_EXPIRE_CACHE_AFTER_WRITE_SECONDS;
+    }
+
+
+    public S3AccessGrantsCachedBucketRegionResolver.Builder toBuilder() {
+        return new S3AccessGrantsCachedBucketRegionResolver.BuilderImpl(this);
+    }
+
+    public static S3AccessGrantsCachedBucketRegionResolver.Builder builder() {
+        return new S3AccessGrantsCachedBucketRegionResolver.BuilderImpl();
+    }
+
+    @Override
+    public Region resolve(String bucket, S3Client s3Client) throws S3Exception {
+
+        this.s3Client = s3Client;
```

Comment:
Why do you need to pass the S3Client as a parameter and then you overwrite the local s3client with the one passed into the function.

Reply:
Was following the same pattern as other cache layer. Passing S3Client can be avoided in this case. Going to be make the changes.